File: examples/go/tests/common_steps.go
package tests

import (
	"context"
	"errors"
	"fmt"
	"strings"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	"github.com/cucumber/godog"
)

// CommonContext holds shared state across scenarios
type CommonContext struct {
	LastError error
}

var sharedContext = &CommonContext{}

// SetLastError stores the last error for assertion in common steps
func SetLastError(err error) {
	sharedContext.LastError = err
}

// InitCommonSteps registers common step definitions shared across aggregates
func InitCommonSteps(ctx *godog.ScenarioContext) {
	// Reset before each scenario
	ctx.Before(func(c context.Context, sc *godog.Scenario) (context.Context, error) {
		sharedContext.LastError = nil
		return c, nil
	})

	// Shared step for checking command failures (used across all aggregates)
	ctx.Step(`^the command fails with status "([^"]*)"$`, commandFailsWithStatus)
	ctx.Step(`^the error message contains "([^"]*)"$`, errorMessageContains)
}

func commandFailsWithStatus(status string) error {
	if sharedContext.LastError == nil {
		return fmt.Errorf("expected command to fail with status %s, but it succeeded", status)
	}
	// Check if it's a CommandRejectedError (value type, not pointer)
	var cmdErr angzarr.CommandRejectedError
	if !errors.As(sharedContext.LastError, &cmdErr) {
		return fmt.Errorf("expected CommandRejectedError, got %T: %v", sharedContext.LastError, sharedContext.LastError)
	}
	return nil
}

func errorMessageContains(text string) error {
	if sharedContext.LastError == nil {
		return fmt.Errorf("expected an error but got success")
	}
	errMsg := strings.ToLower(sharedContext.LastError.Error())
	if !strings.Contains(errMsg, strings.ToLower(text)) {
		return fmt.Errorf("expected error to contain '%s', got '%s'", text, sharedContext.LastError.Error())
	}
	return nil
}

File: examples/cpp/tests/common_steps.cpp
// GTest must be included before cucumber-cpp autodetect for framework detection
#include <gtest/gtest.h>
#include <cucumber-cpp/autodetect.hpp>

#include "test_context.hpp"

using cucumber::ScenarioScope;

/// Before each scenario, reset the context.
BEFORE() { tests::g_context.reset(); }

/// Common step: verify command failed with specific status.
THEN("^the command fails with status \"([^\"]*)\"$") {
    REGEX_PARAM(std::string, expected_status);
    ScenarioScope<tests::ScenarioContext> ctx;

    ASSERT_TRUE(tests::g_context.has_error())
        << "Expected command to fail, but it succeeded";

    auto expected_code = tests::string_to_status_code(expected_status);
    ASSERT_TRUE(tests::g_context.last_error_code.has_value())
        << "Expected error code to be set";

    ASSERT_EQ(tests::g_context.last_error_code.value(), expected_code)
        << "Expected status " << expected_status << " but got "
        << tests::status_code_to_string(tests::g_context.last_error_code.value());
}

/// Common step: verify error message contains substring (case-insensitive).
THEN("^the error message contains \"([^\"]*)\"$") {
    REGEX_PARAM(std::string, expected_substring);

    ASSERT_TRUE(tests::g_context.has_error()) << "Expected command to have failed";
    ASSERT_TRUE(tests::g_context.last_error.has_value()) << "Expected error message to be set";

    // Convert both to lowercase for case-insensitive comparison
    std::string error_msg = tests::g_context.last_error.value();
    std::string expected_lower = expected_substring;
    std::transform(error_msg.begin(), error_msg.end(), error_msg.begin(), ::tolower);
    std::transform(expected_lower.begin(), expected_lower.end(), expected_lower.begin(), ::tolower);

    ASSERT_TRUE(error_msg.find(expected_lower) != std::string::npos)
        << "Expected error message to contain '" << expected_substring << "' but got: '"
        << tests::g_context.last_error.value() << "'";
}
